src/client/visualizations/data-provider/data-provider.tsx

```diff
@@ -0,0 +1,172 @@
+    this.lastQueryEssence = essence;
+    return this.debouncedCallExecutor(essence, timekeeper);
+  }
+
+  protected getQuery(essence: Essence, timekeeper: Timekeeper): Expression {
+    return essence.visualization.name === "grid" ? gridQuery(essence, timekeeper) : visualizationQuery(essence, timekeeper);
+  }
+
+  private callExecutor = (essence: Essence, timekeeper: Timekeeper): Promise<DatasetLoad | null> =>
+    essence.dataCube.executor(this.getQuery(essence, timekeeper), { timezone: essence.timezone })
+      .then((dataset: Dataset) => {
+          // signal out of order requests with null
+          if (!this.wasUsedForLastQuery(essence)) return null;
+          return loaded(dataset);
+        },
+        err => {
+          // signal out of order requests with null
+          if (!this.wasUsedForLastQuery(essence)) return null;
+          reportError(err);
+          return error(err);
+        });
+
+  private wasUsedForLastQuery(essence: Essence) {
+    return essence.equals(this.lastQueryEssence);
+  }
+
+  private debouncedCallExecutor = debounceWithPromise(this.callExecutor, 500);
+
+  private handleDatasetLoad(dataset: DatasetLoad) {
+    this.setState({ dataset });
+    const { registerDownloadableDataset } = this.props;
+    if (registerDownloadableDataset) {
+      registerDownloadableDataset(isLoaded(dataset) ? dataset.dataset : null);
+    }
+  }
+
+  protected shouldFetchData(nextProps: DataProviderProps): boolean {
+    return this.differentVisualizationDefinition(nextProps);
+  }
+
+  private differentVisualizationDefinition(nextProps: DataProviderProps) {
+    const { essence, timekeeper } = this.props;
+    const nextEssence = nextProps.essence;
+    const nextTimekeeper = nextProps.timekeeper;
+    return nextEssence.differentDataCube(essence) ||
+      nextEssence.differentEffectiveFilter(essence, timekeeper, nextTimekeeper) ||
+      nextEssence.differentTimeShift(essence) ||
+      nextEssence.differentSplits(essence) ||
+      nextEssence.differentSeries(essence) ||
+      nextEssence.differentSettings(essence) ||
+      this.differentBucketingTimezone(nextEssence) ||
+      this.differentLastRefreshRequestTimestamp(nextProps);
+  }
+
+  private differentBucketingTimezone(newEssence: Essence): boolean {
+    const { essence } = this.props;
+    return !essence.timezone.equals(newEssence.timezone) && newEssence.splits.hasSplitOn(essence.getTimeDimension());
+  }
+
+  private differentLastRefreshRequestTimestamp({ refreshRequestTimestamp }: DataProviderProps): boolean {
+    return refreshRequestTimestamp !== this.props.refreshRequestTimestamp;
+  }
+
+  private visualisationNotResized(nextProps: DataProviderProps): boolean {
+    return this.props.stage.equals(nextProps.stage);
+  }
+
+  render() {
+    const { children } = this.props;
+    const { dataset } = this.state;
+    switch (dataset.status) {
+      case DatasetLoadStatus.LOADING:
+        return <Loader />;
```

Comment:
Doesn't look good. Loader is shown outside visualisation stage. Absolute positioning?